src/Playwright/Core/RouteHandler.cs

```diff
@@ -88,7 +89,10 @@ public async Task<bool> HandleAsync(Route route)
             Complete = new TaskCompletionSource<bool>(),
             Route = route,
         };
-        _activeInvocations.Add(handlerInvocation);
+        lock (_activeInvocationsLock)
```

Comment:
Did you try with some thread-safe list instead of adding locks?

Reply:
Of course, this can also be solved with concurrent collections, which is also my go-to option for these kinds of problems.
But since we are not in a performance-critical inner loop and the lock is only held for a very short time without user code being called, this is a simple and effective solution with minimal changes to the code.

But I certainly don't insist on this solution. If an alternative solution with the replacement of the type used is desired, then I will do that.